README.md
# QRCode Pix Ruby

[![Maintainability](https://api.codeclimate.com/v1/badges/ff1952ca066c27953773/maintainability)](https://codeclimate.com/github/pedrofurtado/qrcode_pix_ruby/maintainability)
[![codecov](https://codecov.io/gh/pedrofurtado/qrcode_pix_ruby/branch/master/graph/badge.svg?token=OK3S19R5CG)](https://codecov.io/gh/pedrofurtado/qrcode_pix_ruby)
[![CI](https://github.com/pedrofurtado/qrcode_pix_ruby/actions/workflows/ci.yml/badge.svg)](https://github.com/pedrofurtado/qrcode_pix_ruby/actions/workflows/ci.yml)
[![Gem Version](https://badge.fury.io/rb/qrcode_pix_ruby.svg)](https://badge.fury.io/rb/qrcode_pix_ruby)
[![Gem](https://img.shields.io/gem/dt/qrcode_pix_ruby.svg)]()
[![license](https://img.shields.io/github/license/pedrofurtado/qrcode_pix_ruby.svg)]()
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://github.com/pedrofurtado/qrcode_pix_ruby)
[![Ruby Style Guide](https://img.shields.io/badge/code_style-rubocop-brightgreen.svg)](https://github.com/rubocop/rubocop)

Ruby gem for Qrcode generation of Pix (Pagamento Instantâneo Brasileiro - Banco Central do Brasil).

<img style="max-width: 100%;" src="https://github.com/pedrofurtado/qrcode_pix_ruby/blob/master/pix_logo.png?raw=true" height="100px" />

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'qrcode_pix_ruby'
```

And then execute:

    bundle install

Or install it yourself as:

    gem install qrcode_pix_ruby

## Usage (examples)

### Available parameters

- pix_key/url
- repeatable
- currency
- country_code
- description
- postal_code
- merchant_name
- merchant_city
- transaction_id
- amount

### Static QR Code

```ruby
require 'qrcode_pix_ruby'

pix = QrcodePixRuby::Payload.new(
  pix_key:        'minhachavedopix',
  description:    'Pagamento do pedido 123456',
  merchant_name:  'Fulano de Tal',
  merchant_city:  'SAO PAULO',
  transaction_id: 'TID12345',
  amount:         '100.00',
  currency:       '986',
  country_code:   'BR',
  postal_code:    '01131010',
  repeatable:     false
)

# If needed, change the attributes value later
pix.pix_key = 'minhaoutrachavepix'
pix.merchant_city = 'BRASILIA'

# QRCode copia-e-cola
puts pix.payload

# QRCode for images
puts pix.base64
```

### Dynamic QR Code

```ruby
require 'qrcode_pix_ruby'

pix = QrcodePixRuby::Payload.new(
  url:            'https://example.com',
  merchant_name:  'Fulano de Tal',
  merchant_city:  'SAO PAULO',
  amount:         '100.00',
  transaction_id: 'TID12345',
  repeatable:     false
)

# If needed, change the attributes value later
pix.url = 'https://another-example.com'
pix.amount = 1.50

# QRCode copia-e-cola
puts pix.payload

# QRCode para uso em imagens
puts pix.base64
```

🚨 Important note: BACEN (Banco Central do Brasil) sets a variety of rules for each field in QRCode Payload (maximum number of characters, invalid values, mandatory and optional fields, and so on). So, pay attention that a QRCode payload+base64 generated by the gem can be valid, theorically, but may not be accepted in banking apps because of these restrictions (that are out of scope of gem).

## Execute tests/specs locally

To execute gem tests locally, use Docker with the commands below:

```bash
git clone https://github.com/pedrofurtado/qrcode_pix_ruby
cd qrcode_pix_ruby/
docker build -t qrcode_pix_ruby_specs .

# Then, run this command how many times you want,
# after editing local files, and so on, to get
# feedback from test suite of gem.
docker run -v $(pwd):/app/ -it qrcode_pix_ruby_specs
```

## Execute demo locally

To execute demo locally, use Docker with the commands below:

```bash
git clone https://github.com/pedrofurtado/qrcode_pix_ruby
cd qrcode_pix_ruby/demo/
docker build -t qrcode_pix_ruby_demo .

# Then, access http://localhost:3000 the see demo in action.
docker run -p 3000:3000 -it qrcode_pix_ruby_demo
```

## Useful links

* https://github.com/joseviniciusnunes/qrcode-pix
* https://www.bcb.gov.br/content/estabilidadefinanceira/forumpireunioes/AnexoI-PadroesParaIniciacaodoPix.pdf
* https://github.com/renatomb/php_qrcode_pix
* https://www.gerarpix.com.br
* https://github.com/fbbergamo/gerador-pix
* https://pix.ae
* https://pix.nascent.com.br/tools/pix-qr-decoder
* https://openpix.com.br/qrcode/scanner
* https://openpix.com.br/qrcode/debug
* https://github.com/william-costa/wdev-qrcode-pix-estatico-php
* https://github.com/william-costa/wdev-qrcode-pix-php
* https://www.youtube.com/watch?v=eO11iFgrdCA
* https://qrcodepix.dinheiro.tech
* http://decoder.qrcodepix.dinheiro.tech
* https://www.bcb.gov.br/estabilidadefinanceira/pix
* https://gerencianet.com.br/blog/qr-code-estatico-qr-code-dinamico-no-pix
* https://blog.juno.com.br/pix-qr-code-estatico-x-qr-code-dinamico
* https://github.com/entria/awesome-pix
* https://zxing.org/w/decode.jspx
* https://github.com/hiagodotme/gpix
* https://github.com/EnssureIT/faz-um-pix
* https://github.com/jesobreira/pixkey
* https://github.com/pedrinholemes/pix-br
* https://github.com/ceciliadeveza/gerarqrcodepix
* https://www.bcb.gov.br/content/estabilidadefinanceira/SiteAssets/Manual%20do%20BR%20Code.pdf
* https://www.bcb.gov.br/content/estabilidadefinanceira/spb_docs/ManualBRCode.pdf

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/pedrofurtado/qrcode_pix_ruby. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/pedrofurtado/qrcode_pix_ruby/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the qrcode_pix_ruby project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/pedrofurtado/qrcode_pix_ruby/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

- ...

## [0.6.0]- 2023-04-14

- Fixed CI
- Fixed CRC16 algorithm with rjust with 4 digits

## [0.5.0] - 2021-07-13

- Add option to pass Pix informations directly in initializer
- CI speedup
- Specs reorganization

## [0.4.0] - 2021-07-10

- Add URL option for PIX, in order to generate dynamic QR Codes

## [0.3.4] - 2021-06-28

- Fix behavior of EMV field 'repeatable' (the logic was inversed)

## [0.3.3] - 2021-06-25

- Rubocop setup
- Add rubies 2.3, 2.4, 2.5 and 2.6 to support of gem

## [0.3.2] - 2021-06-24

- Setup for run tests/specs locally with Docker

## [0.3.1] - 2021-06-24

- More documentation

## [0.1.0...0.3.0] - 2021-06-24

- Initial release
- Setup of CI and integrations
- Skeleton of gem


spec/spec_helper.rb
# frozen_string_literal: true

if ENV['CODECOV_TOKEN']
  require 'simplecov'
  SimpleCov.start
  require 'codecov'
  SimpleCov.formatter = SimpleCov::Formatter::Codecov
end

require 'qrcode_pix_ruby'

Dir['./spec/support/**/*.rb'].sort.each { |f| require f }

RSpec.configure do |config|
  config.example_status_persistence_file_path = '.rspec_status'
  config.disable_monkey_patching!
  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/pix_ruby_spec.rb
# frozen_string_literal: true

RSpec.describe QrcodePixRuby do
  it 'has a valid version number' do
    expect(QrcodePixRuby::VERSION).not_to be nil
    expect(QrcodePixRuby::VERSION).to match /^[0-9]+\.[0-9]+\.[0-9]+$/
  end

  it 'generates default payload and base64, when nothing is defined' do
    pix = QrcodePixRuby::Payload.new
    expect(pix.payload).to eq '00020126260014br.gov.bcb.pix0100020052040000530398654005802BR5900600062040500630462E0'
    expect(pix.base64).to eq ''
  end

  it 'generates valids payloads and base64' do
    pix = QrcodePixRuby::Payload.new

    pix.pix_key        = '12345678900'
    pix.description    = 'Pagamento do pedido 123456'
    pix.merchant_name  = 'William Costa'
    pix.merchant_city  = 'SAO PAULO'
    pix.transaction_id = 'WDEV1234'
    pix.amount         = '100.00'

    expect(pix.payload).to eq '00020126630014br.gov.bcb.pix0111123456789000226Pagamento do pedido 1234565204000053039865406100.005802BR5913William Costa6009SAO PAULO62120508WDEV12346304E9BF'
    expect(pix.base64).to eq ''
  end
end


spec/qrcode_pix_ruby_spec.rb
# frozen_string_literal: true

RSpec.describe QrcodePixRuby do
  let(:attributes) { {} }

  let(:payload) { subject.payload }
  let(:base64) { subject.base64 }

  describe 'Attributes validation' do
    subject { QrcodePixRuby::Payload.new(**attributes) }

    let(:error) { QrcodePixRuby::PayloadArgumentError }

    context 'only one' do
      let(:attributes) { { abc: 'abc' } }

      it 'raise error and show the unknown attribute' do
        expect { subject }.to raise_error(error, 'Unknown attributes: abc')
      end
    end

    context 'more then one' do
      let(:attributes) { { abc: 'abc', dfg: 'dfg' } }

      it 'raise error and show the unknown attributes' do
        expect { subject }.to raise_error(error, 'Unknown attributes: abc, dfg')
      end
    end
  end

  describe 'Static QR Code' do
    subject do
      QrcodePixRuby::Payload.new(
        pix_key: 'pedro.felipe.azevedo.furtado@gmail.com',
        merchant_name: 'PEDRO FURTADO',
        merchant_city: 'SAO PAULO',
        **attributes
      )
    end

    describe 'QRCode - Basic' do
      it 'has a valid payload' do
        expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865802BR5913PEDRO FURTADO6009SAO PAULO62070503***63046B4C'
      end

      it 'has a valid base64' do
        expect(base64).to eq file_fixture('basic.base64').read
      end
    end

    describe 'QRCode - Currency' do
      context 'when currency is 986' do
        let(:attributes) { { currency: '986' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865802BR5913PEDRO FURTADO6009SAO PAULO62070503***63046B4C'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('currency_986.base64').read
        end
      end

      context 'when currency is 123' do
        let(:attributes) { { currency: '123' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053031235802BR5913PEDRO FURTADO6009SAO PAULO62070503***63047E2C'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('currency_123.base64').read
        end
      end
    end

    describe 'QRCode - Amount' do
      context 'when amount is 100.99' do
        let(:attributes) { { amount: '100.99' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865406100.995802BR5913PEDRO FURTADO6009SAO PAULO62070503***63045059'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('amount_100_99.base64').read
        end
      end

      context 'when amount is generate a small crc16 checksum' do
        let(:attributes) { { amount: '1.02' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com52040000530398654041.025802BR5913PEDRO FURTADO6009SAO PAULO62070503***630409C3'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('amount_1_02.base64').read
        end
      end

      context 'when amount is 0.01' do
        let(:attributes) { { amount: '0.01' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com52040000530398654040.015802BR5913PEDRO FURTADO6009SAO PAULO62070503***6304A136'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('amount_0_01.base64').read
        end
      end
    end

    describe 'QRCode - Country code' do
      context 'when country_code is BR' do
        let(:attributes) { { country_code: 'BR' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865802BR5913PEDRO FURTADO6009SAO PAULO62070503***63046B4C'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('country_code_BR.base64').read
        end
      end

      context 'when country_code is US' do
        let(:attributes) { { country_code: 'US' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865802US5913PEDRO FURTADO6009SAO PAULO62070503***63049DF5'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('country_code_US.base64').read
        end
      end
    end

    describe 'QRCode - Postal code' do
      context 'when postal_code is 01134000' do
        let(:attributes) { { postal_code: '01134000' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865802BR5913PEDRO FURTADO6009SAO PAULO61080113400062070503***630497E2'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('postal_code_01134000.base64').read
        end
      end

      context 'when postal_code is 01131010' do
        let(:attributes) { { postal_code: '01131010' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865802BR5913PEDRO FURTADO6009SAO PAULO61080113101062070503***63041100'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('postal_code_01131010.base64').read
        end
      end
    end

    describe 'QRCode - Transaction ID' do
      context 'when transaction_id is PIX123' do
        let(:attributes) { { transaction_id: 'PIX123' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865802BR5913PEDRO FURTADO6009SAO PAULO62100506PIX12363046BE4'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('transaction_id_PIX123.base64').read
        end
      end

      context 'when transaction_id is PIXPEDRO' do
        let(:attributes) { { transaction_id: 'PIXPEDRO' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865802BR5913PEDRO FURTADO6009SAO PAULO62120508PIXPEDRO630428E7'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('transaction_id_PIXPEDRO.base64').read
        end
      end
    end

    describe 'QRCode - Description' do
      context 'when description is Bora pagar?' do
        let(:attributes) { { description: 'Bora pagar?' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226750014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com0211Bora pagar?5204000053039865802BR5913PEDRO FURTADO6009SAO PAULO62070503***6304BC58'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('description_Bora_pagar.base64').read
        end
      end

      context 'when description is PIX 123' do
        let(:attributes) { { description: 'PIX 123' } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226710014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com0207PIX 1235204000053039865802BR5913PEDRO FURTADO6009SAO PAULO62070503***6304D936'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('description_PIX_123.base64').read
        end
      end
    end

    describe 'QRCode - Repeatable' do
      context 'when repeatable is false' do
        let(:attributes) { { repeatable: false } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021226600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865802BR5913PEDRO FURTADO6009SAO PAULO62070503***63046B4C'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('repeatable_false.base64').read
        end
      end

      context 'when repeatable is true' do
        let(:attributes) { { repeatable: true } }

        it 'has a valid payload' do
          expect(payload).to eq '00020101021126600014BR.GOV.BCB.PIX0138pedro.felipe.azevedo.furtado@gmail.com5204000053039865802BR5913PEDRO FURTADO6009SAO PAULO62070503***630468A4'
        end

        it 'has a valid base64' do
          expect(base64).to eq file_fixture('repeatable_true.base64').read
        end
      end
    end
  end

  describe 'Dynamic QR Code' do
    subject do
      QrcodePixRuby::Payload.new(
        url: 'https://qrcodes-pix-h.gerencianet.com.br/v2/3074067216ec4341949e07872a163799',
        merchant_name: 'William Costa',
        merchant_city: 'SAO PAULO',
        amount: '10.00',
        transaction_id: 'WDEV1234123412340000000005',
        repeatable: false,
        **attributes
      )
    end

    describe 'QRCode - URL' do
      it 'has a valid payload' do
        expect(payload).to eq '00020101021226900014BR.GOV.BCB.PIX2568qrcodes-pix-h.gerencianet.com.br/v2/3074067216ec4341949e07872a163799520400005303986540510.005802BR5913William Costa6009SAO PAULO62300526WDEV12341234123400000000056304D37C'
      end

      it 'has a valid base64' do
        expect(base64).to eq file_fixture('url.base64').read
      end
    end
  end
end


lib/pix_ruby/qr_code/static.rb
# frozen_string_literal: true

module PixRuby
  module QrCode
    class Static
      ID_PAYLOAD_FORMAT_INDICATOR = '00'.freeze
      ID_MERCHANT_ACCOUNT_INFORMATION = '26'.freeze
      ID_MERCHANT_ACCOUNT_INFORMATION_GUI = '00'.freeze
      ID_MERCHANT_ACCOUNT_INFORMATION_KEY = '01'.freeze
      ID_MERCHANT_ACCOUNT_INFORMATION_DESCRIPTION = '02'.freeze
      ID_MERCHANT_CATEGORY_CODE = '52'.freeze
      ID_TRANSACTION_CURRENCY = '53'.freeze
      ID_TRANSACTION_AMOUNT = '54'.freeze
      ID_COUNTRY_CODE = '58'.freeze
      ID_MERCHANT_NAME = '59'.freeze
      ID_MERCHANT_CITY = '60'.freeze
      ID_ADDITIONAL_DATA_FIELD_TEMPLATE = '62'.freeze
      ID_ADDITIONAL_DATA_FIELD_TEMPLATE_TXID = '05'.freeze
      ID_CRC16 = '63'.freeze

      attr_accessor :pix_key, :description, :merchant_name, :merchant_city, :txid, :amount

      def payload
        p  = get_value(ID_PAYLOAD_FORMAT_INDICATOR, '01')
        p += get_merchant_account_information
        p += get_value(ID_MERCHANT_CATEGORY_CODE, '0000')
        p += get_value(ID_TRANSACTION_CURRENCY, '986')
        p += get_value(ID_TRANSACTION_AMOUNT, amount)
        p += get_value(ID_COUNTRY_CODE, 'BR')
        p += get_value(ID_MERCHANT_NAME, merchant_name)
        p += get_value(ID_MERCHANT_CITY, merchant_city)
        p += get_additional_data_field_template

        p + crc16(p)
      end

      private

      def get_merchant_account_information
        # dominio do banco
        gui = get_value(ID_MERCHANT_ACCOUNT_INFORMATION_GUI, 'br.gov.bcb.pix')

        #chave pix
        key = get_value(ID_MERCHANT_ACCOUNT_INFORMATION_KEY, pix_key)

        #descrição do pagamento
        payment_text = get_value(ID_MERCHANT_ACCOUNT_INFORMATION_DESCRIPTION, description)

        # valor completo da conta
        get_value ID_MERCHANT_ACCOUNT_INFORMATION, "#{gui}#{key}#{payment_text}"
      end

      def get_value(id, value)
        size = value.to_s.length.to_s.rjust(2, '0')
        "#{id}#{size}#{value}"
      end

      def get_additional_data_field_template
        #txid
        calculated_txid = get_value ID_ADDITIONAL_DATA_FIELD_TEMPLATE_TXID, txid
        get_value ID_ADDITIONAL_DATA_FIELD_TEMPLATE, calculated_txid
      end

      def crc16(raw_payload)
        p = "#{raw_payload}#{ID_CRC16}04"

        polinomio = 0x1021
        resultado = 0xFFFF

        length = p.length
        if length > 0
          offset = 0
          while offset < length
            resultado = resultado ^ (p[offset].bytes[0] << 8)

            bitwise = 0
            while bitwise < 8
              resultado = resultado << 1

              if resultado & 0x10000 != 0
                resultado = resultado ^ polinomio
              end

              resultado = resultado & 0xFFFF

              bitwise += 1
            end

            offset += 1
          end
        end

        "#{ID_CRC16}04#{resultado.to_s(16).upcase}"
      end
    end
  end
end


lib/qrcode_pix_ruby/payload.rb
# frozen_string_literal: true

require 'rqrcode'

module QrcodePixRuby
  class PayloadArgumentError < ArgumentError; end

  class Payload
    ID_PAYLOAD_FORMAT_INDICATOR                 = '00'
    ID_POINT_OF_INITIATION_METHOD               = '01'
    ID_MERCHANT_ACCOUNT_INFORMATION             = '26'
    ID_MERCHANT_ACCOUNT_INFORMATION_GUI         = '00'
    ID_MERCHANT_ACCOUNT_INFORMATION_KEY         = '01'
    ID_MERCHANT_ACCOUNT_INFORMATION_DESCRIPTION = '02'
    ID_MERCHANT_ACCOUNT_INFORMATION_URL         = '25'
    ID_MERCHANT_CATEGORY_CODE                   = '52'
    ID_TRANSACTION_CURRENCY                     = '53'
    ID_TRANSACTION_AMOUNT                       = '54'
    ID_COUNTRY_CODE                             = '58'
    ID_MERCHANT_NAME                            = '59'
    ID_MERCHANT_CITY                            = '60'
    ID_POSTAL_CODE                              = '61'
    ID_ADDITIONAL_DATA_FIELD_TEMPLATE           = '62'
    ID_ADDITIONAL_DATA_FIELD_TEMPLATE_TXID      = '05'
    ID_CRC16                                    = '63'

    ATTRIBUTES = %i[pix_key url repeatable currency country_code description postal_code
                    merchant_name merchant_city transaction_id amount].freeze

    def initialize(**kwargs)
      verify_kwargs(kwargs.keys)

      ATTRIBUTES.each do |attribute|
        singleton_class.class_eval { attr_accessor attribute }
        instance_variable_set("@#{attribute}", kwargs[attribute])
      end
    end

    def payload
      p = ''

      p += emv(ID_PAYLOAD_FORMAT_INDICATOR, '01')
      p += emv_repeatable
      p += emv_merchant
      p += emv(ID_MERCHANT_CATEGORY_CODE, '0000')
      p += emv(ID_TRANSACTION_CURRENCY, currency || '986')
      p += emv(ID_TRANSACTION_AMOUNT, amount) if amount
      p += emv(ID_COUNTRY_CODE, country_code || 'BR')
      p += emv(ID_MERCHANT_NAME, merchant_name)
      p += emv(ID_MERCHANT_CITY, merchant_city)
      p += emv(ID_POSTAL_CODE, postal_code) if postal_code
      p += emv_additional_data

      p + crc16(p)
    end

    def base64
      ::RQRCode::QRCode.new(payload).as_png(
        bit_depth: 1,
        border_modules: 0,
        color_mode: 0,
        color: 'black',
        file: nil,
        fill: 'white',
        module_px_size: 6,
        resize_exactly_to: false,
        resize_gte_to: false
      ).to_data_url
    end

    private

    def verify_kwargs(keys)
      unknowns = keys - ATTRIBUTES

      return unless unknowns.any?

      raise QrcodePixRuby::PayloadArgumentError, "Unknown attributes: #{unknowns.join(', ')}"
    end

    def emv(id, value)
      size = value.to_s.length.to_s.rjust(2, '0')
      "#{id}#{size}#{value}"
    end

    def emv_repeatable
      emv(ID_POINT_OF_INITIATION_METHOD, repeatable ? '11' : '12')
    end

    def emv_merchant
      merchant_gui         = emv(ID_MERCHANT_ACCOUNT_INFORMATION_GUI, 'BR.GOV.BCB.PIX')
      merchant_pix_key     = emv(ID_MERCHANT_ACCOUNT_INFORMATION_KEY, pix_key) if pix_key
      merchant_description = emv(ID_MERCHANT_ACCOUNT_INFORMATION_DESCRIPTION, description) if description
      merchant_url         = emv(ID_MERCHANT_ACCOUNT_INFORMATION_URL, url.gsub(/https?:\/\//, '')) if url
      emv(ID_MERCHANT_ACCOUNT_INFORMATION, "#{merchant_gui}#{merchant_pix_key}#{merchant_description}#{merchant_url}")
    end

    def emv_additional_data
      txid = emv(ID_ADDITIONAL_DATA_FIELD_TEMPLATE_TXID, transaction_id || '***')
      emv(ID_ADDITIONAL_DATA_FIELD_TEMPLATE, txid)
    end

    def crc16(text)
      extended_payload = "#{text}#{ID_CRC16}04"
      extended_payload_length = extended_payload.length
      polynomial = 0x1021
      result = 0xFFFF

      if extended_payload_length.positive?
        offset = 0

        while offset < extended_payload_length
          result ^= extended_payload[offset].bytes[0] << 8
          bitwise = 0

          while bitwise < 8
            result <<= 1
            result ^= polynomial if result & 0x10000 != 0
            result &= 0xFFFF
            bitwise += 1
          end

          offset += 1
        end
      end

      "#{ID_CRC16}04#{result.to_s(16).upcase.rjust(4, '0')}"
    end
  end
end
